config/dynamic.c

```diff
@@ -16,11 +16,546 @@
+#define LUA_DY_BASE_DIR     "./lua"
+#define DYNAMIC_CONFIG_FILE "dylib.lua"
+
+/*
+ * The cast "(uint16_t)" is needed because of
+ * the strict compilation check of DPDK,
+ * without uint16_t, it pops an error message
+ * "error: large integer implicitly truncated to
+ * unsigned type [-Werror=overflow]."
+ */
+static const uint16_t MSG_MAX_LEN = (uint16_t)~0U;
+
+static struct dynamic_config config;
+
+/*
+ * Return values:
+ * 	0: Write @nbytes successfully.
+ * 	-1: Connection closed by the client.
+ */
+static int
+write_nbytes(int conn_fd, const char *msg_buff, int nbytes)
+{
+	int send_size;
+	int tot_size = 0;
+
+	if (nbytes == 0)
+		return 0;
+
+	while ((send_size = write(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += send_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the client is closed. */
+	if (send_size == 0) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"dyn_cfg: Client disconnected\n");
+		return -1;
+	}
+
+	if (send_size < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to write data to the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+reply_client_message(int conn_fd, const char *reply_msg, uint16_t reply_len)
+{
+	int ret;
+	uint16_t nlen = htons(reply_len);
+
+	/* The first two bytes: the length of the message in network order. */
+	ret = write_nbytes(conn_fd, (char *)&nlen, sizeof(nlen));
+	if (ret != 0)
+		return -1;
+
+	/* Sending the message. */
+	ret = write_nbytes(conn_fd, reply_msg, reply_len);
+	if (ret != 0)
+		return -1;
+
+	return 0;
+}
+
+static int
+process_client_message(int conn_fd,
+	const char *msg, int msg_len, lua_State *lua_state)
+{
+	/*
+	 * TODO Implement the functionalities to process the clients' request.
+	 *
+	 * Gatekeeper: Adding, listing, and removing BP components;
+	 *
+	 * Gatekeeper: Adding, listing, removing IP ranges
+	 * handled by the global LPM table in GKs;
+	 *
+	 * Gatekeeper and Grantor: Listing the ARP and ND table.
+	 * This is important for network diagnosis.
+	 *
+	 * Grantor: Updating the policy GTs run.
+	 */
+
+	int ret;
+	size_t reply_len;
+	const char *reply_msg;
+	const char *CLIENT_PROC_ERROR =
+		"function reply() return -1, \"Dynamic Configuration cannot process the request. end\"\n";
+
+	if (msg_len == 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: the received message is an empty string\n"); 
+		return -1;
+	}
+
+	/* Load the client's Lua chunk, and run it. */
+	ret = luaL_loadbuffer(lua_state, msg, msg_len, "message")
+		|| lua_pcall(lua_state, 0, 1, 0);
+	if (ret != 0) {
+		reply_msg = lua_tostring(lua_state, -1);
+
+		RTE_LOG(ERR, LUA, "dyn_cfg: %s!\n", reply_msg);
+
+		reply_len = strlen(reply_msg);
+		if (reply_len > MSG_MAX_LEN) {
+			RTE_LOG(WARNING, LUA,
+				"dyn_cfg: The error message length (%lu) exceeds the limit!\n",
+				reply_len);
+			reply_len = MSG_MAX_LEN;
+		}
+
+		reply_client_message(conn_fd, reply_msg, reply_len);
+
+		return -1;
```

Comment:
Combine the two lines above as follows:
`return reply_client_message(conn_fd, reply_msg, reply_len);`

There's no reason to abort the connection since we can gracefully handle this case.